elyra/pipeline/component.py

```diff
@@ -252,13 +260,57 @@ def resource_type(self):
         return self.location_type
 
     @abstractmethod
-    def read_component_definition(self, location: str) -> Optional[str]:
+    def read_component_definition(self,
+                                  location: str,
+                                  location_to_def: Dict[str, str]) -> Dict[str, str]:
         """
         Read an absolute location to get the contents of a component
         specification file
+
+        :param location: an absolute path to the specification file to read
+        :param location_to_def: a mapping of component locations to definitions
+
+        :returns: the given 'location_to_def' object, optionally including a new
+                  key-value pair if the given component location is successfully read
         """
         raise NotImplementedError()
 
+    def read_component_definitions(self, locations: List[str]) -> Dict[str, str]:
+        """
+        This function starts 3 or fewer threads that read component definitions in parallel.
+
+        The 'location_to_def' variable is a mapping of a component location to its content.
+        As a mutable object, this dictionary provides a means to retrieve a return value for
+        each thread. If a thread is able to successfully read the content of the given
+        component file location, a location-to-content mapping is added to 'location_to_def'.
+        """
+        location_to_def = {}
+
+        loc_q = Queue()
+        for location in self.get_absolute_locations(locations):
+            loc_q.put_nowait(location)
+
+        def read_with_thread():
+            """Read the contents of a given location from the queue"""
+            while not loc_q.empty():
+                try:
+                    loc = loc_q.get(timeout=.1)
+                except Empty:
+                    continue
+                self.read_component_definition(loc, location_to_def)
```

Comment:
After thinking about this more and since we have a fixed set of threads, we should protect the thread from unwarranted termination by wrapping `self.read_component_definition()` in a `try/except`. 

We should also test this with a "bad edit" where the edit to the registry is for an unavailable or inaccessible URL.  Seems like we need to deal with "poison messages" in some manner.